src/roles/dullahan.py

```diff
@@ -103,100 +100,88 @@ def on_del_player(evt, cli, var, nick, nickrole, nicktpls, death_triggers):
 
+@event_listener("swap_player")
+def on_swap(evt, var, old_user, user):
+    if old_user in KILLS:
+        KILLS[user] = KILLS.pop(old_user)
+    if old_user in TARGETS:
+        TARGETS[user] = TARGETS.pop(old_user)
+
+    for dullahan, target in KILLS.items():
+        if target is old_user:
+            KILLS[dullahan] = user
+
+    for dullahan, targets in TARGETS.items():
+        if old_user in targets:
+            targets.remove(old_user)
+            targets.add(user)
+
 @event_listener("get_special")
 def on_get_special(evt, cli, var):
     evt.data["special"].update(var.ROLES["dullahan"])
 
 @event_listener("transition_day", priority=2)
 def on_transition_day(evt, cli, var):
-    for k, d in list(KILLS.items()):
-        evt.data["victims"].append(d)
-        evt.data["onlybywolves"].discard(d)
-        evt.data["killers"][d].append(k)
-        del KILLS[k]
+    while KILLS:
+        k, d = KILLS.popitem()
+        evt.data["victims"].append(d.nick)
+        evt.data["onlybywolves"].discard(d.nick)
+        evt.data["killers"][d.nick].append(k.nick)
 
 @event_listener("exchange_roles")
 def on_exchange(evt, cli, var, actor, nick, actor_role, nick_role):
-    if actor in KILLS:
-        del KILLS[actor]
-    if nick in KILLS:
-        del KILLS[nick]
-    if actor_role == "dullahan" and nick_role != "dullahan" and actor in TARGETS:
-        TARGETS[nick] = TARGETS[actor] - {nick}
-        del TARGETS[actor]
-    elif nick_role == "dullahan" and actor_role != "dullahan" and nick in TARGETS:
-        TARGETS[actor] = TARGETS[nick] - {actor}
-        del TARGETS[nick]
+    for k in set(KILLS):
+        if k.nick == actor or k.nick == nick:
+            del KILLS[k]
+
+    for k in set(TARGETS):
+        if actor_role == "dullahan" and nick_role != "dullahan" and k.nick == actor:
+            TARGETS[users._get(nick)] = TARGETS.pop(k) - {users._get(nick)} # FIXME
+        elif nick_role == "dullahan" and actor_role != "dullahan" and k.nick == nick:
+            TARGET[users._get(actor)] = TARGETS.pop(k) - {users._get(actor)} # FIXME
 
 @event_listener("chk_nightdone")
 def on_chk_nightdone(evt, cli, var):
-    spl = set(list_players())
+    spl = set(get_players())
     evt.data["actedcount"] += len(KILLS)
-    for p in var.ROLES["dullahan"]:
-        if TARGETS[p] & spl:
-            evt.data["nightroles"].append(p)
+    for d, targets in TARGETS.items():
+        if targets & spl:
+            evt.data["nightroles"].append(d.nick)
 
 @event_listener("transition_night_end", priority=2)
 def on_transition_night_end(evt, cli, var):
-    for dullahan in var.ROLES["dullahan"]:
-        targets = list(TARGETS[dullahan])
-        for target in var.DEAD:
-            if target in targets:
-                targets.remove(target)
-        if not targets: # already all dead
-            pm(cli, dullahan, "{0} {1}".format(messages["dullahan_simple"], messages["dullahan_targets_dead"]))
+    for dullahan, targets in TARGETS.items():
```

Comment:
Keep this iterating over var.ROLES["dullahan"] -- we want this to break if someone manages to be a dullahan without having an entry in TARGETS. Not doing this would result in silent and harder-to-track-down bugs (chk_nightdone is fine as-is though, don't need to change that back).